src/util/pg.py

```diff
@@ -196,27 +196,56 @@ def explode_query_range(cr, query, table, alias=None, bucket_size=10000, prefix=
 
     alias = alias or table
 
-    cr.execute("SELECT min(id), max(id) FROM {}".format(table))
-    min_id, max_id = cr.fetchone()
-    if min_id is None:
-        return []  # empty table
-
     if "{parallel_filter}" not in query:
         sep_kw = " AND " if re.search(r"\sWHERE\s", query, re.M | re.I) else " WHERE "
         query += sep_kw + "{parallel_filter}"
 
-    if ((max_id - min_id + 1) * 0.9) <= bucket_size:
-        # If there is less than `bucket_size` records (with a 10% tolerance), no need to explode the query.
-        # Force usage of `prefix` in the query to validate it correctness.
-        # If we don't the query may only be valid if there is no split. It avoid scripts to pass the CI but fail in production.
+    cr.execute(format_query(cr, "SELECT min(id), max(id) FROM {}", table))
+    min_id, max_id = cr.fetchone()
+    if min_id is None:
+        return []  # empty table
+    ids = list(range(min_id, max_id + 1, bucket_size))
+    if len(ids) > 100000:
+        _logger.getChild("explode_query_range").warning(
+            "Refusing to run %s queries, switching to a precise bucketing strategy", len(ids) - 1
+        )
+        cr.execute(
+            format_query(
+                cr,
+                """
+                WITH t AS (
+                    SELECT id,
+                           mod(row_number() OVER(ORDER BY id) - 1, %s) AS g
+                      FROM {table}
+                     ORDER BY id
+                ) SELECT array_agg(id) FILTER (WHERE g=0),
+                         min(id),
+                         max(id)
+                    FROM t
+                """,
+                table=table,
+            ),
+            [bucket_size],
+        )
+        ids, min_id, max_id = cr.fetchone()
+        if min_id is None:
+            return []  # empty table
+
+    assert min_id == ids[0] and max_id + 1 != ids[-1]  # sanity checks
```

Comment:
```suggestion
                         max(id)
                    FROM t
                """,
                table=table,
            ),
            [bucket_size],
        )
        ids, max_id = cr.fetchone()
        if max_id is None:
            return []  # empty table

    min_id = ids[0]
```
I don't think we need to assume bugs in PostgreSQL, here.

Reply:
It's not PG whom I doubt, it's _us_. This will server as a check point if anything done a few lines above is adapted in the future ;)